pkg/pool/pool.go

```diff
@@ -359,26 +441,45 @@ func (p *pool) ReconcilePool() *worker.WarmPoolJob {
 		p.pendingCreate += deviation
 
 		log.Info("created job to add resources to warm pool", "requested count", deviation)
-
+		if p.isPDPool {
+			return worker.NewWarmPoolCreateJob(p.nodeName, deviation/NumIPv4AddrPerPrefix)
+		}
 		return worker.NewWarmPoolCreateJob(p.nodeName, deviation)
 
 	} else if -deviation > p.warmPoolConfig.MaxDeviation {
 		// Need to delete from warm pool
 		deviation = -deviation
+
 		var resourceToDelete []string
-		for i := len(p.warmResources) - 1; i >= len(p.warmResources)-deviation; i-- {
-			resourceToDelete = append(resourceToDelete, p.warmResources[i])
+		numToDelete := deviation
+		for numToDelete > 0 {
+			numFreeGroups, freeResourceGroups := p.findFreeGroup()
+			if numFreeGroups == 0 {
+				log.Info("no warm resources to delete", "deviation", deviation, "numToDelete", numToDelete)
+				break
+			}
+
+			// Remove resources to be deleted form the warm pool
+			for _, groupID := range freeResourceGroups {
+				if numToDelete <= 0 {
+					break
+				}
+				log.Info("removing free resource group from warm pool", "group id", groupID)
+				resourceToDelete = append(resourceToDelete, groupID)
+				numToDelete -= len(p.warmResources[groupID])
+				delete(p.warmResources, groupID)
+			}
 		}
 
-		// Remove resources to be deleted form the warm pool
-		p.warmResources = p.warmResources[:len(p.warmResources)-deviation]
 		// Increment pending to the number of resource being deleted, once successfully deleted the count can be decremented
-		p.pendingDelete += deviation
-		// Submit the job to delete resources
+		p.pendingDelete += deviation - numToDelete
```

Comment:
Shouldn't `numToDelete` be 0 here?

Reply:
Yea you are right here. Updated accordingly.